### PR TITLE
Hibernate with Panache: rework getEntityManager and flush

### LINKED ISSUE #14542: static flush or automaticlly flush static methods for hibernate reactive panache
**Description**
I tried to figure out why my PanacheEntityBase::deleteById not actually delete anything in the database with the debugger, and found out that the DeleteEvent only queued but never executed. Then I found out that there's a PanacheEntityBase::flush that can flush the change to the database. But the flush method is not static, so I had to write the following code
`deleteById(id).call(ignore -> new MyEntity().flush());` which is weird. Also the javadoc of deleteById and such didn't tell me that the action need flush.

**Implementation ideas**
There can be three approachs to solve this, only I don't know which is better:
1. make PanacheEntityBase::flush static
2. add a static variant of PanacheEntityBase::flush, which require one of them to rename
3. invoke flush in all the static method changes database, such ad deleteById and so on

### LINKED ISSUE #15595: DB/Panache API around getting sessions is badly designed
In ORM with Panache we have:

- `PanacheEntityBase.getEntityManager()` which is wrongly an instance method, it should be static since it applies to entity classes, never instances.
- `PanacheEntityBase.flush()` is similarly wrongly an instance method because it flushes the entire DB, not just one entity. It's convenient, sure, but deceptive.
- Probably we're missing `Panache.flush()` as a shortcut for `Panache.getEntityManager().flush()` too, in case we have a single PersistenceUnit.

In HR with Panache we're missing:

- `PanacheEntityBase.getSession()` for an equivalent to `getEntityManager()`
- Probably we're missing `Panache.flush()` as a shortcut for `Panache.getSession().flush()` too, in case we have a single PersistenceUnit.

Do we have similar issues in Mongo/Panache @loicmathieu?

I'm not sure how to deal with the change from instance to static though, given that it's not binary compatible at all. Adding them with worse names doesn't sound very appealing.

### PR DESCRIPTION
**Pull request by loicmathieu**
Fixes #15595
Fixes #14542

### CODE REVIEW

**Comment by loicmathieu** [comment]
@FroMage before implementing the Kotlin and reactive side I wanted to have your feedback on this one:
- I remove the `getEntityManager(Class<?>)` method from PanacheRepository as PanacheEntity and PanacheRepository should only give access to the entity manager of the current entity
- I had to move the `Panache` class, but as it's on the same package it should be fine, but I would need to duplicate it for Kotlin users :(
- I'm not fan of the methods on `Panache` that takes the class to retrieve the right entity manager, the class is used to retrive the name of the persistance unit, maybe those methods should take a string as parameter ?

**Comment by FroMage** [comment]
> I remove the getEntityManager(Class<?>) method from PanacheRepository as PanacheEntity and PanacheRepository should only give access to the entity manager of the current entity

+1

> I had to move the Panache class, but as it's on the same package it should be fine, but I would need to duplicate it for Kotlin users :(

Can you explain why you had to move it?

> I'm not fan of the methods on `Panache` that takes the class to retrieve the right entity manager, the class is used to retrive the name of the persistance unit, maybe those methods should take a string as parameter ?

I can see only one: `getEntityManager(Class)` and it makes a lot of sense to me, since that's the only place to be able to obtain an EM for a given class. I do agree we could add a version that takes a `String persistenceUnit`.

In fact, `executeUpdate` and `flush` would need to have some sort of equivalent for specifying an alternate PU, but so far nobody asked for it.

When people ask for it, we can introduce a `PanacheOperations` interface with the same operations as instance methods, and add `PanacheOperations Panache.for(Class/String)` which would return the operations class for the given PU, and `PanacheOperations Panache.default()` for the default PU one that the current static methods point to?

**Comment by loicmathieu** [comment]
@FroMage, I added reactive, kotlin and update what's needed in data-rest and spring-data.

> Can you explain why you had to move it?

Because it needs to access the `JpaOperations` that is in the another extension.

> I can see only one: getEntityManager(Class) and it makes a lot of sense to me, since that's the only place to be able to obtain an EM for a given class. I do agree we could add a version that takes a String persistenceUnit.

OK, I keep the `Class` variant and added a `String` variant. Maybe the easiest way will be to replace those two variants with a variant that take the `EntityManager` ? This can also be added to the existing executeUpdate methods without too many overload.

> When people ask for it, we can introduce a PanacheOperations interface 

This is interesting, we can then add more opreations on this `PanacheOperations` (people often ask for some convenience database methods, we can put them here later) if needed and deprecate the one directly on Panache.

**Comment by FroMage** [comment]
> OK, I keep the Class variant and added a String variant. Maybe the easiest way will be to replace those two variants with a variant that take the EntityManager ? This can also be added to the existing executeUpdate methods without too many overload.

Well the point of those methods is to make it easy to access operations without getting hold of an EM, so I guess I'd rather not.

**Review by FroMage — DISMISSED** [review]
Changes look good, but we should add tests for the new methods.

**Comment by FroMage** [comment]
Those failing tests don't look good though ;)

**Comment by loicmathieu** [comment]
@FroMage I add back the getEntityManager(Class) methods on the repository interface and make them deprecated to avoid too many breaking changes. Can you validate the PR ?

@evanchooly can you have a quick look at the kotlin side?

**Review by evanchooly — APPROVED** [review]
In general, it looks good to me.  there appears to be a mix of calling static methods on subtypes which isn't my favorite but it works.

File: extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java [hunk 1]
```diff
@@ -39,7 +51,7 @@ public static EntityManager getEntityManager(Class<?> clazz) {
      * @return the current {@link TransactionManager}
      */
     public static TransactionManager getTransactionManager() {
-        return AbstractJpaOperations.getTransactionManager();
+        return JpaOperations.getTransactionManager();
```
[the remark this answers is not in the document]
Comment (loicmathieu) [reply extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/Panache.java:54]: Yes it's static.
I'll revert the change.

**Comment by loicmathieu** [comment]
@FroMage I added tests as you requested and the failing test are in an unrelated module, is it OK for you ? Can we merge it ?

**Review by FroMage — COMMENTED** [review]
LGTM, thanks!

**Pull request merged**